File: HNLineAnimationTextField-OC/HNLineAnimationTextFieldDemo-OC/HNLineAnimationTextField/Categories/UIView+HNHierarchy.m
//
//  UIView+HNHierarchy.m
//  HNLineAnimationTextFieldDemo-OC
//
//  Created by zakariyyaSv on 16/4/24.
//  Copyright © 2016年 Zakariyya. All rights reserved.
//

#import "UIView+HNHierarchy.h"
#import <objc/runtime.h>
#import "NSArray+HNSort.h"

@implementation UIView (HNHierarchy)

//Special textFields,textViews,scrollViews
Class UIAlertSheetTextFieldClass;       //UIAlertView
Class UIAlertSheetTextFieldClass_iOS8;  //UIAlertView

Class UITableViewCellScrollViewClass;   //UITableViewCell
Class UITableViewWrapperViewClass;      //UITableViewCell
Class UIQueuingScrollViewClass;         //UIPageViewController

Class UISearchBarTextFieldClass;        //UISearchBar

+ (void)initialize{
    [super initialize];
    UIAlertSheetTextFieldClass          = NSClassFromString(@"UIAlertSheetTextField");
    UIAlertSheetTextFieldClass_iOS8     = NSClassFromString(@"_UIAlertControllerTextField");
    
    UITableViewCellScrollViewClass      = NSClassFromString(@"UITableViewCellScrollView");
    UITableViewWrapperViewClass         = NSClassFromString(@"UITableViewWrapperView");
    UIQueuingScrollViewClass            = NSClassFromString(@"_UIQueuingScrollView");
    
    UISearchBarTextFieldClass           = NSClassFromString(@"UISearchBarTextField");

}

- (void)_setIsAskingCanBecomeFirstResponder:(BOOL)isAskingCanBecomeFirstResponder{
    objc_setAssociatedObject(self, @selector(isAskingCanBecomeFirstResponder), @(isAskingCanBecomeFirstResponder), OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}

- (BOOL)isAskingCanBecomeFirstResponder{
    NSNumber *isAskingCanBecomeFirstResponder = objc_getAssociatedObject(self, @selector(isAskingCanBecomeFirstResponder));
    return [isAskingCanBecomeFirstResponder boolValue];
}

// return view's controller
- (UIViewController *)viewController{
    UIResponder *nextResponder = self;
    do {
        nextResponder = [nextResponder nextResponder];
        if ([nextResponder isKindOfClass:[UIViewController class]]) {
            return (UIViewController *)nextResponder;
        }
    } while (nextResponder != nil);
    return nil;
}

- (UIViewController *)topMostController{
    NSMutableArray *controllersHierarchy = [[NSMutableArray alloc] init];
    UIViewController *topController = self.window.rootViewController;
    if (topController) {
        [controllersHierarchy addObject:topController];
    }
    while ([topController presentedViewController]) {
        topController = [topController presentedViewController];
        [controllersHierarchy addObject:topController];
    }
    UIResponder *matchController = [self viewController];
    while (matchController != nil && ![controllersHierarchy containsObject:matchController]) {
        do {
            matchController = [matchController nextResponder];
        } while (matchController != nil && ![matchController isKindOfClass:[UIViewController class]]);
    }
    return (UIViewController *)matchController;
}

- (BOOL)HNcanBecomeFirstResponder{
    [self _setIsAskingCanBecomeFirstResponder:YES];
    BOOL _HNcanBecomeFirstResponder = ([self canBecomeFirstResponder] && [self isUserInteractionEnabled] && ![self isHidden] && self.alpha != 0);
    if (_HNcanBecomeFirstResponder == YES) {
        if ([self isKindOfClass:[UITextField class]]) {
            _HNcanBecomeFirstResponder = [(UITextField *)self isEnabled];
        }
    }
    [self _setIsAskingCanBecomeFirstResponder:NO];
    return _HNcanBecomeFirstResponder;
}

// get all UITextfields that can become first responder
- (NSArray *)responderSiblings{
    NSArray *siblings = self.superview.subviews;
    NSMutableArray *textFields = [[NSMutableArray alloc] init];
    for (UITextField *textfield in siblings) {
        if ([textfield HNcanBecomeFirstResponder]) {
            [textFields addObject:textfield];
        }
    }
    return textFields;
}

- (NSArray *)deepResponderViews{
    NSMutableArray *textfields = [[NSMutableArray alloc] init];
    NSArray *subviews = [self.subviews positionArr];
    for (UITextField *textfield in subviews) {
        if ([textfield HNcanBecomeFirstResponder]) {
            [textfields addObject:textfield];
        }
        else if (textfield.subviews.count && [textfield isUserInteractionEnabled] && ![textfield isHidden] && [textfield alpha] != 0){
            [textfields addObjectsFromArray:[textfield deepResponderViews]];
        }
    }
    return textfields;
}

// get view hierarchy depth
- (NSInteger)depth{
    NSInteger depth = 0;
    if ([self superview]) {
        depth = [[self superview] depth] + 1;
    }
    return depth;
}

- (NSString *)subHierarchy{
    NSMutableString *debugInfo = [[NSMutableString alloc] initWithString:@"\n"];
    NSInteger depth = [self depth];
    for (int counter = 0; counter < depth; counter++) {
        [debugInfo appendString:[self debugHierarchy]];
    }
    for (UIView *subView in self.subviews) {
        [debugInfo appendString:[subView debugHierarchy]];
    }
    return debugInfo;
}

- (NSString *)superHierarchy
{
    NSMutableString *debugInfo = [[NSMutableString alloc] init];
    
    if (self.superview)
    {
        [debugInfo appendString:[self.superview superHierarchy]];
    }
    else
    {
        [debugInfo appendString:@"\n"];
    }
    
    NSInteger depth = [self depth];
    
    for (int counter = 0; counter < depth; counter ++)  [debugInfo appendString:@"|  "];
    
    [debugInfo appendString:[self debugHierarchy]];
    
    [debugInfo appendString:@"\n"];
    
    return debugInfo;
}

-(NSString *)debugHierarchy
{
    NSMutableString *debugInfo = [[NSMutableString alloc] init];
    
    [debugInfo appendFormat:@"%@: ( %.0f, %.0f, %.0f, %.0f )",NSStringFromClass([self class]), CGRectGetMinX(self.frame), CGRectGetMinY(self.frame), CGRectGetWidth(self.frame), CGRectGetHeight(self.frame)];
    
    if ([self isKindOfClass:[UIScrollView class]])
    {
        UIScrollView *scrollView = (UIScrollView*)self;
        [debugInfo appendFormat:@"%@: ( %.0f, %.0f )",NSStringFromSelector(@selector(contentSize)),scrollView.contentSize.width,scrollView.contentSize.height];
    }
    
    if (CGAffineTransformEqualToTransform(self.transform, CGAffineTransformIdentity) == false)
    {
        [debugInfo appendFormat:@"%@: %@",NSStringFromSelector(@selector(transform)),NSStringFromCGAffineTransform(self.transform)];
    }
    
    return debugInfo;
}

- (BOOL)isSearchBarTextField{
    return ([self isKindOfClass:UISearchBarTextFieldClass] || [self isKindOfClass:[UISearchBar class]]);
}

- (BOOL)isAlertViewTextField{
    return ([self isKindOfClass:UIAlertSheetTextFieldClass] || [self isKindOfClass:UIAlertSheetTextFieldClass_iOS8]);
}

- (BOOL)isTextView{
    return ([self isKindOfClass:[UITextView class]]);
}

@end

@implementation NSObject (HN_logging)

- (NSString *)HNDescription{
    return [NSString stringWithFormat:@"<%@,%p>",NSStringFromClass([self class]),self];
}

@end

File: HNLineAnimationTextFieldDemo-swift/HNLineAnimationTextFieldDemo-swift/HNLineAnimationTextField/UIView+HNHierarchy.swift
//
//  UIView+HNHierarchy.swift
//  HNLineAnimationTextFieldDemo-swift
//
//  Created by zakariyyaSv on 16/4/26.
//  Copyright © 2016年 Zakariyya. All rights reserved.
//

import Foundation
import UIKit

private var HNIsAskingCanBecomeFirstResponder = "HNIsAskingCanBecomeFirstResponder"

public extension UIView {
    
    public var isAskingCanBecomeFirstResponder: Bool {
        get {
            if  let value = objc_getAssociatedObject(self, &HNIsAskingCanBecomeFirstResponder) as? Bool {
                return value
            }
            else{
                return false
            }
        }
        set(newValue) {
            objc_setAssociatedObject(self, &HNIsAskingCanBecomeFirstResponder, newValue,objc_AssociationPolicy.OBJC_ASSOCIATION_RETAIN_NONATOMIC)
        }
    }
    
    public func viewController() -> UIViewController? {
        var nextResponder : UIResponder? = self
        repeat {
            nextResponder = nextResponder?.nextResponder()
            if let viewController = nextResponder as? UIViewController {
                return viewController
            }
        }while nextResponder != nil
        return nil
    }
    
    public func topMostController() -> UIViewController? {
        var controllerHierarchy = [UIViewController]()
        if var topController = window?.rootViewController {
            controllerHierarchy.append(topController)
            while topController.presentedViewController != nil {
                topController = topController.presentedViewController!
                controllerHierarchy.append(topController)
            }
            var matchController : UIResponder? = viewController()
            
            while matchController != nil && controllerHierarchy.contains(matchController as! UIViewController) == false {
                repeat {
                    matchController = matchController?.nextResponder()
                }while matchController != nil && matchController is UIViewController == false
            }
            return matchController as? UIViewController
        }
        else{
            return viewController()
        }
        
    }
    
    // Returns all siblings that canBecomeFirstResponder
    public func responderSiblings() -> [UIView] {
        var tempTextFields = [UIView]()
        if let siblings = superview?.subviews {
            for textfield in siblings {
                if textfield._HNcanBecomeFirstResponder() == true {
                    tempTextFields.append(textfield)
                }
            }
        }
        return tempTextFields
    }
    
    public func deepResponderViews() -> [UIView] {
        let subViews = subviews.sort { (obj1 : AnyObject, obj2 : AnyObject) -> Bool in
            let view1 = obj1 as! UIView
            let view2 = obj2 as! UIView
            
            let x1 = CGRectGetMinX(view1.frame)
            let y1 = CGRectGetMinY(view1.frame)
            let x2 = CGRectGetMinX(view2.frame)
            let y2 = CGRectGetMinY(view2.frame)
            
            if y1 != y2 {
                return y1 < y2
            }
            else{
                return x1 < x2
            }
        }
        var textfields = [UIView]()
        
        for textfield in subViews {
            if textfield._HNcanBecomeFirstResponder() == true {
                textfields.append(textfield)
            }
            else if textfield.subviews.count != 0 && userInteractionEnabled == true && hidden == false && alpha != 0.0 {
                for deepView in textfield.deepResponderViews() {
                    textfields.append(deepView)
                }
            }
        }
        
        return textfields
    }
    
    private func _HNcanBecomeFirstResponder() -> Bool{
        isAskingCanBecomeFirstResponder = true
        var _HNCanBecomeFirstResponder = (canBecomeFirstResponder() == true && userInteractionEnabled == true && hidden == false && alpha != 0.0 && isAlertViewTextField() == false && isSearchBarTextField() == false)
        if _HNCanBecomeFirstResponder == true {
            if let textfield = self as? UITextField {
                _HNCanBecomeFirstResponder = textfield.enabled
            }
        }
        isAskingCanBecomeFirstResponder = false
        return _HNCanBecomeFirstResponder
    }
    
    public func isAlertViewTextField() -> Bool{
        struct InternalClass {
            static var UIAlertSheetTextFieldClass : AnyClass? = NSClassFromString("UIAlertSheetTextField")
            static var UIAlertSheetTextFieldClass_iOS8 : AnyClass? = NSClassFromString("_UIAlertControllerTextField")
        }
        return (InternalClass.UIAlertSheetTextFieldClass != nil && isKindOfClass(InternalClass.UIAlertSheetTextFieldClass!)) || (InternalClass.UIAlertSheetTextFieldClass_iOS8 != nil && isKindOfClass(InternalClass.UIAlertSheetTextFieldClass_iOS8!))
    }
    
    public func isSearchBarTextField() -> Bool {
        struct InternalClass {
            static var UISearchBarTextFieldClass: AnyClass? = NSClassFromString("UISearchBarTextField")
        }
        
        return (InternalClass.UISearchBarTextFieldClass != nil && isKindOfClass(InternalClass.UISearchBarTextFieldClass!)) || self is UISearchBar
    }
    
    public func isTextView() -> Bool {
        return isKindOfClass(UITextView)
    }
    
    private func depth() -> Int {
        var depth : Int = 0
        if let superView = superview {
            depth = superView.depth() + 1
        }
        return depth
        
    }
    
}

extension NSObject {
    public func HNDescription() -> String {
        return "<\(self) \(unsafeAddressOf(self))>"
    }
}